keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,165 @@
+Support memory qos in pod qos class using cgroups v2.
+
+## Motivation
+In traditional cgroups v1 implement in Kubernetes, pod qos class and qos manager can only limit cpu resources, such as cpu_shares, memory qos is not yet implemented. cgroups v2 brings new capabilities for memory controller and it would help Kubernetes enhance memory isolation quality.
+
+### Goals
+- Support memory qos for pod qos classes
+
+### Non-Goals
+- additional qos design
+- support other resources qos
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` memory allocation overcommit ( request < limit ) could increase OOM kill possibility because of node memory running out
+
+Currently we only use `memory.limit_in_bytes == limits.memory` to limit memory hardly in cgroups v1 and `memory.max == limits.memory` in cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request memory for `Guaranteed` pods. And for `Burstable` pods, it uses `memory.low` to achieve best-effort memory protection as well as using `memory.high` to throttle allocation of memory overcommit when pods specify `requests.memory`. We also set `memory.max = limits.memory` to do hard limit to make OOM killer work.
```

Comment:
Maybe we can set `memory.min = requests.memory` in burstable pods as same as guaranteed to protect memory uses, and throttle the allocation above `requests.memory`? Any ideas @bobbypage

Reply:
hmm, what do you mean by "and throttle the allocation above requests.memory", how would that work?

This is a good question, I would be interested to further understand what's pros / cons of using memory.min for memory request as opposed to memory.low for burstable workloads and what side affects this could have.

Reply:
> what do you mean by "and throttle the allocation above requests.memory", how would that work?
That is because we also set `memory.high = requests.memory` of burstable pods, and if the pod's memory goes over the `memory.high` boundary, it would be throttled and put under high reclaim pressure.

Reply:
> memory.high = requests.memory of burstable pods

I see, thanks for explanation. Thinking about that a bit more, I'm not sure if it makes sense to set `memory.high = requests.memory` for burstable pods. From user perspective, in cgroup v1 world, memory request is declaring how much memory the application needs to be scheduled and doesn't say anything about the max limit. 

As an example, you can create a burstable pod with some memory request and no limit set today, and that pods would be able to use all available memory on system (which is a good thing from utilization perspective!). By setting memory.high == memory.request, this is putting artificial memory cap which breaks the utilization aspect and also may be surprising to users if suddenly their pods would start hitting memory reclaim when they their usage was >= request.

Reply:
Yes you are right. `memory.high  = requests.memory` may be more aggressive and cause confusion for users.

Reply:
Done.